[AutoDiff] Type-checking support for `inout` parameter differentiation.

dan-zheng opened the pull request:
Semantically, an `inout` parameter is both a parameter and a result.

`@differentiable` and `@derivative` attributes now support original functions
with one "semantic result": either a formal result or an `inout` parameter.

The differential/pullback type of a function with `inout` differentiability
parameters also has `inout` parameters. This is ideal for performance.

Differential typing rules:
- Case 1: original function has no `inout` parameters.
  - Original:     `(T0, T1, ...) -> R`
  - Differential: `(T0.Tan, T1.Tan, ...) -> R.Tan`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> Void`
  - Differential: `(T0.Tan, ...) -> T1.Tan`
- Case 3: original function has a wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> Void`
  - Differential: `(T0.Tan, inout T1.Tan, ...) -> Void`

Pullback typing rules:
- Case 1: original function has no `inout` parameters.
  - Original: `(T0, T1, ...) -> R`
  - Pullback: `R.Tan -> (T0.Tan, T1.Tan, ...)`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> Void`
  - Pullback: `(T1.Tan) -> (T0.Tan, ...)`
- Case 3: original function has a wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> Void`
  - Pullback: `(inout T1.Tan) -> (T0.Tan, ...)`

Resolves TF-1164.

---

Examples:

```swift
extension Float {
  @derivative(of: *=)
  static func _vjpMultiplyAssign(_ lhs: inout Float, _ rhs: Float) -> (
    value: Void, pullback: (inout Float) -> Float
  ) {
    defer { lhs *= rhs }
    return ((), { [lhs = lhs] v in
      let drhs = lhs * v
      v *= rhs
      return drhs
    })
  }
}

extension Array where Element: Differentiable {
  @derivative(of: append)
  mutating func _vjpAppend(_ element: Element) -> (
    value: Void, pullback: (inout TangentVector) -> Element.TangentVector
  ) {
    let appendedElementIndex = count
    defer { append(element) }
    return ((), { dself in dself.base[appendedElementIndex] })
  }
}
```

---

This patch includes all `inout` parameter differentiation changes currently
relevant to `master` branch, e.g. SIL derivative type calculation changes.

Remaining changes involves code that hasn't yet been upstreamed. Those changes
will be committed to `tensorflow` branch and eventually upstreamed.

End-to-end `inout` parameter differentiation is tested in https://github.com/apple/swift/pull/29956.

dan-zheng committed:
[AutoDiff] Type-checking support for `inout` parameter differentiation.

Semantically, an `inout` parameter is both a parameter and a result.

`@differentiable` and `@derivative` attributes now support original functions
with one "semantic result": either a formal result type or an `inout` parameter.

The differential/pullback type of a function with `inout` differentiability
parameters also has `inout` parameters. This is ideal for performance.

Differential typing rules:
- Case 1: original function has no `inout` parameters.
  - Original:     `(T0, T1, ...) -> R`
  - Differential: `(T0.Tan, T1.Tan, ...) -> R.Tan`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> Void`
  - Differential: `(T0.Tan, ...) -> T1.Tan`
- Case 3: original function has a wrt `inout` parameter.
  - Original:     `(T0, inout T1, ...) -> R`
  - Differential: `(T0.Tan, inout T1.Tan, ...) -> Void`

Pullback typing rules:
- Case 1: original function has no `inout` parameters.
  - Original: `(T0, T1, ...) -> R`
  - Pullback: `R.Tan -> (T0.Tan, T1.Tan, ...)`
- Case 2: original function has a non-wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> Void`
  - Pullback: `(T1.Tan) -> (T0.Tan, ...)`
- Case 3: original function has a wrt `inout` parameter.
  - Original: `(T0, inout T1, ...) -> R`
  - Pullback: `(inout T1.Tan) -> (T0.Tan, ...)`

Resolves TF-1164.

dan-zheng commented:
Let's check that tests pass.
@swift-ci Please smoke test

dan-zheng commented on include/swift/AST/Types.h:
```diff
@@ -3267,12 +3282,55 @@ class AnyFunctionType : public TypeBase {
+  /// first. `makeSelfParamFirst` should be true when working with user-facing
+  /// derivative function types, e.g. when type-checking `@differentiable` and
+  /// `@derivative` attributes.
+  AnyFunctionType *getAutoDiffReturnedLinearMapFunctionType(
```
Note: this helper exists for code sharing between `AnyFunctionType::getAutoDiffDerivativeFunctionType` and `@derivative` attribute type-checking.

Perhaps `getAutoDiffDerivativeFunctionLinearMapResultType` is a better name.

dan-zheng replied:
Renamed to `getAutoDiffDerivativeFunctionLinearMapResultType`.

dan-zheng commented on test/AutoDiff/Sema/differentiable_attr_type_checking.swift:
```diff
@@ -1158,6 +1146,42 @@ final class FinalClass: Differentiable {
 
+// Test `inout` parameters.
+
+@differentiable(wrt: y)
```
Note: I intentionally didn't add tests for `@differentiable(jvp:vjp)`, since that's actively being removed (TF-1162).

dan-zheng commented:
@swift-ci Please smoke test

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -113,6 +113,13 @@ struct AutoDiffConfig {
 
+/// A semantic function result type: either a formal function result type or
+/// an `inout` parameter. Used in derivative function type calculation.
+struct SemanticFunctionResultType {
```
Should make this name be more specific to AD, e.g. `AutoDiffSemanticFunctionResultType`.

dan-zheng replied:
Good point. Done.

rxwei commented on include/swift/AST/Types.h:
```diff
@@ -3267,12 +3282,55 @@ class AnyFunctionType : public TypeBase {
+  /// first. `makeSelfParamFirst` should be true when working with user-facing
+  /// derivative function types, e.g. when type-checking `@differentiable` and
+  /// `@derivative` attributes.
+  AnyFunctionType *getAutoDiffDerivativeFunctionLinearMapResultType(
```
```suggestion
  AnyFunctionType *getAutoDiffDerivativeFunctionLinearMapType(
```
To make it clear it's returning the type of the linear map, not the linear map's result.

dan-zheng replied:
That is better. Done.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3849,6 +3839,7 @@ bool checkIfDifferentiableProgrammingEnabled(
 
 llvm::Expected<IndexSubset *> DifferentiableAttributeTypeCheckRequest::evaluate(
     Evaluator &evaluator, DifferentiableAttr *attr) const {
+#if 0
```
what are all the `#if 0`s for?

dan-zheng replied:
Oops, those changes were not intentional. I've reverted them, the latest commit just addresses your review feedback.

dan-zheng committed:
Address review feedback.

- `SemanticFunctionResultType` -> `AutoDiffSemanticFunctionResultType`
- `getAutoDiffDerivativeFunctionLinearMapResultType` ->
  `getAutoDiffDerivativeFunctionLinearMapType`

rxwei commented on lib/AST/AutoDiff.cpp:
```diff
@@ -99,11 +98,56 @@ void autodiff::getSubsetParameterTypes(IndexSubset *subset,
 }
 
+AutoDiffSemanticFunctionResultType
+autodiff::getOriginalFunctionSemanticResultType(
```
What's the word "original" trying to disambiguate?  Seems redundant to me.

dan-zheng replied:
> What's the word "original" trying to disambiguate? Seems redundant to me.

Thanks for the catch! I renamed `getOriginalFunctionSemanticResultType` to `getFunctionSemanticResultType` and simplified it.

dan-zheng committed:
Simplify `getFunctionSemanticResultType`.

Rename `getOriginalFunctionSemanticResultType` to `getFunctionSemanticResultType`.

It is no longer strictly related to "original" functions: scrub all mentions of
"original".

rxwei commented on include/swift/AST/Types.h:
```diff
@@ -3267,12 +3282,55 @@ class AnyFunctionType : public TypeBase {
+  ///   - Original: `(T0, inout T1, ...) -> Void`
+  ///   - Pullback: `(T1.Tan) -> (T0.Tan, ...)`
+  /// - Case 3: original function has a wrt `inout` parameter.
+  ///   - Original: `(T0, inout T1, ...) -> R`
```
```suggestion
  ///   - Original: `(T0, inout T1, ...) -> Void`
```

rxwei commented on lib/AST/Type.cpp:
```diff
@@ -5028,6 +4969,109 @@ AnyFunctionType *AnyFunctionType::getAutoDiffDerivativeFunctionType(
+    // - Pullback: `(T1.Tan) -> (T0.Tan, ...)`
+    //
+    // Case 3: original function has a wrt `inout` parameter.
+    // - Original: `(T0, inout T1, ...) -> R`
```
```suggestion
    // - Original: `(T0, inout T1, ...) -> Void`
```

rxwei commented on lib/AST/Type.cpp:
```diff
@@ -5028,6 +4969,109 @@ AnyFunctionType *AnyFunctionType::getAutoDiffDerivativeFunctionType(
+    // - Differential: `(T0.Tan, ...) -> T1.Tan`
+    //
+    // Case 3: original function has a wrt `inout` parameter.
+    // - Original:     `(T0, inout T1, ...) -> R`
```
```suggestion
    // - Original:     `(T0, inout T1, ...) -> Void`
```

rxwei commented:
> Case 3: original function has a wrt `inout` parameter.
>   * Original:     `(T0, inout T1, ...) -> R`
>   * Differential: `(T0.Tan, inout T1.Tan, ...) -> Void`
> Case 3: original function has a wrt `inout` parameter.
> * Original: `(T0, inout T1, ...) -> R`
> * Pullback: `(inout T1.Tan) -> (T0.Tan, ...)`

I don't think these are correct. `R` should be a `Void` because those the wrt `inout` parameter already behaves as a result. There cannot be a second result.

rxwei commented:
> ```swift
> static func _vjpMultiplyAssign(_ lhs: inout Float, _ rhs: ${Self}) -> (
> ```

There are multiple occurrences of `${Self}` in your code example in the PR description.

rxwei commented on include/swift/AST/Types.h:
```diff
@@ -3267,12 +3282,55 @@ class AnyFunctionType : public TypeBase {
       GenericSignature derivativeGenericSignature = GenericSignature(),
       bool makeSelfParamFirst = false);
 
+  /// Returns the linear map function type returned by the derivative function
```
The documentation seems unclear. Does this expect `this` to be "the derivative function", or any function? If you intend to use normal (original) functions as `this`, then the "returned by the derivative function" part is confusing -- the concept of "linear map functions of a function" doesn't have to be associated with "derivative functions". I'd just say "Returns the corresponding linear map function type for the given ...".

dan-zheng replied:
Nice tip! Done:
```
  /// Returns the corresponding linear map function type for the given parameter
  /// indices, linear map function kind, and other auxiliary parameters.
```

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -262,11 +269,19 @@ using SILDifferentiabilityWitnessKey = std::pair<StringRef, AutoDiffConfig>;
+/// Remaps the original semantic result using `genericEnv`, if specified.
+AutoDiffSemanticFunctionResultType
+getFunctionSemanticResultType(AnyFunctionType *functionType,
+                              bool &hasMultipleSemanticResults,
```
This argument is not being used by any caller, and AD isn't going to support multiple semantic results any time soon.

rxwei replied:
Nvm, I'm seeing a user now. But this function still seems problematic in that when there are multiple semantic results it is unclear which result the returned value is.  Why not, instead, define this function as:

```swift
void getFunctionSemanticResultTypes(
    AnyFunctionType *functionType,
    GenericEnvironment *genericEnv = nullptr,
    SmallVectorImpl<AutoDiffSemanticFunctionResultType> &result)
```

dan-zheng replied:
`getFunctionSemanticResultTypes` is indeed nicer than `getFunctionSemanticResultType` with the `hasMultipleSemanticResults` boolean.

Done:
```cpp
/// Given a function type, collects its semantic result types in type order
/// into `result`: first, the formal result type (if non-`Void`), followed by
/// `inout` parameter types.
///
/// The function type may have at most two parameter lists.
///
/// Remaps the original semantic result using `genericEnv`, if specified.
void getFunctionSemanticResultTypes(
    AnyFunctionType *functionType,
    SmallVectorImpl<AutoDiffSemanticFunctionResultType> &result,
    GenericEnvironment *genericEnv = nullptr);
```

rxwei commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3034,6 +3032,11 @@ ERROR(autodiff_attr_original_decl_none_valid_found,none,
+ERROR(autodiff_attr_original_void_result,none,
+      "cannot differentiate void function %0", (DeclName))
+ERROR(autodiff_attr_original_multiple_semantic_results,none,
+      "cannot yet differentiate functions with more than one semantic result "
```
- "Semantic result" isn't really a user-facing term until we added the whole "inout differentiation" chapter to the manifesto (and the language book in the future)... Something more specific to `inout` would be better.
    > cannot differentiate functions with both an 'inout' parameter and a result
- Since differentiating multiple semantic results is pretty far away, I'd suggest just saying "cannot" instead of "cannot yet".

dan-zheng replied:
Done: the new message is `cannot differentiate functions with both an 'inout' parameter and a result`.

dan-zheng committed:
Address review feedback.

dan-zheng commented on include/swift/AST/Types.h:
```diff
@@ -3267,12 +3282,55 @@ class AnyFunctionType : public TypeBase {
+  ///   - Original:     `(T0, inout T1, ...) -> Void`
+  ///   - Differential: `(T0.Tan, ...) -> T1.Tan`
+  /// - Case 3: original function has a wrt `inout` parameter.
+  ///   - Original:     `(T0, inout T1, ...) -> R`
```
Oops, thanks for the catch! Fixed.

dan-zheng commented on include/swift/AST/Types.h:
```diff
@@ -3257,6 +3267,11 @@ class AnyFunctionType : public TypeBase {
   ///
+  /// The original type may have `inout` parameters. If so, the
+  /// differential/pullback typing rules are more nuanced: see documentation for
+  /// `getAutoDiffReturnedLinearMapFunctionType` for details. Semantically,
```
Done.

dan-zheng commented:
@swift-ci Please smoke test
